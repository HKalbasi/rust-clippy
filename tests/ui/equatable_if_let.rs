// run-rustfix

#![allow(unused_variables, dead_code, clippy::redundant_pattern_matching, clippy::op_ref)]
#![warn(clippy::equatable_if_let, clippy::equatable_matches)]

use std::cmp::Ordering;

#[derive(PartialEq)]
enum Enum {
    TupleVariant(i32, u64),
    RecordVariant { a: i64, b: u32 },
    UnitVariant,
    Recursive(Struct),
}

#[derive(PartialEq)]
struct Struct {
    a: i32,
    b: bool,
}

#[derive(Clone, Copy)]
enum NotPartialEq {
    A,
    B,
}

#[derive(Clone, Copy)]
enum NotStructuralEq {
    A,
    B,
}

impl PartialEq for NotStructuralEq {
    fn eq(&self, _: &NotStructuralEq) -> bool {
        false
    }
}

#[derive(PartialEq)]
enum Generic<A, B> {
    VA(A),
    VB(B),
    VC,
}

#[derive(PartialEq)]
struct Generic2<A, B> {
    a: A,
    b: B,
}

fn macro_pattern() {
    macro_rules! m1 {
        (x) => {
            "abc"
        };
    }
    if let m1!(x) = "abc" {
        println!("OK");
    }
}

fn main() {
    let a = 2;
    let b = 3;
    let c = Some(2);
    let d = Struct { a: 2, b: false };
    let e = Enum::UnitVariant;
    let f = NotPartialEq::A;
    let g = NotStructuralEq::A;
    let h: Generic<Enum, NotPartialEq> = Generic::VC;
    let i: Generic<Enum, NotStructuralEq> = Generic::VC;
    let j = vec![1, 2, 3, 4];
    let k = Some(&false);
    let l = Generic2 {
        a: Generic2 { a: "xxxx", b: 3 },
        b: Generic2 {
            a: &Enum::UnitVariant,
            b: false,
        },
    };
    let m = Generic2 { a: 3, b: 5 };
    let n = Some("xxxx");
    let mut o = j.iter();

    // true

    if let 2 = a {}
    if let "hello" = "world" {}
    if let Ordering::Greater = a.cmp(&b) {}
    if let Enum::UnitVariant = e {}
    if let None = Some(g) {}
    if let Generic::VC = i {}
    if let None = k {}

    let _ = matches!(b, 2);

    // false

    if let Some(2) = c {}
    if let Struct { a: 2, b: false } = d {}
    if let Enum::TupleVariant(32, 64) = e {}
    if let Enum::RecordVariant { a: 64, b: 32 } = e {}
    if let Some("yyy") = n {}

    let _ = matches!(c, Some(2));

    while let Some(4 | 7) = o.next() {}
}
