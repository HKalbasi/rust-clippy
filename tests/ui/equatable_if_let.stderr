error: this pattern matching can be expressed using equality
  --> $DIR/equatable_if_let.rs:59:8
   |
LL |     if let m1!(x) = "abc" {
   |        ^^^^^^^^^^^^^^^^^^ help: try: `"abc" == m1!(x)`
   |
   = note: `-D clippy::equatable-if-let` implied by `-D warnings`

error: this pattern matching can be expressed using equality
  --> $DIR/equatable_if_let.rs:89:8
   |
LL |     if let 2 = a {}
   |        ^^^^^^^^^ help: try: `a == 2`

error: this pattern matching can be expressed using equality
  --> $DIR/equatable_if_let.rs:90:8
   |
LL |     if let "hello" = "world" {}
   |        ^^^^^^^^^^^^^^^^^^^^^ help: try: `"world" == "hello"`

error: this pattern matching can be expressed using equality
  --> $DIR/equatable_if_let.rs:91:8
   |
LL |     if let Ordering::Greater = a.cmp(&b) {}
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `a.cmp(&b) == Ordering::Greater`

error: this pattern matching can be expressed using equality
  --> $DIR/equatable_if_let.rs:92:8
   |
LL |     if let Enum::UnitVariant = e {}
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `e == Enum::UnitVariant`

error: this pattern matching can be expressed using equality
  --> $DIR/equatable_if_let.rs:93:8
   |
LL |     if let None = Some(g) {}
   |        ^^^^^^^^^^^^^^^^^^ help: try: `Some(g) == None`

error: this pattern matching can be expressed using equality
  --> $DIR/equatable_if_let.rs:94:8
   |
LL |     if let Generic::VC = i {}
   |        ^^^^^^^^^^^^^^^^^^^ help: try: `i == Generic::VC`

error: this pattern matching can be expressed using equality
  --> $DIR/equatable_if_let.rs:95:8
   |
LL |     if let None = k {}
   |        ^^^^^^^^^^^^ help: try: `k == None`

error: this pattern matching can be expressed using equality
  --> $DIR/equatable_if_let.rs:97:13
   |
LL |     let _ = matches!(b, 2);
   |             ^^^^^^^^^^^^^^ help: try: `b == 2`
   |
   = note: `-D clippy::equatable-matches` implied by `-D warnings`

error: aborting due to 9 previous errors

